src/coreclr/jit/lowerxarch.cpp

```diff
@@ -1398,6 +1398,27 @@ GenTree* Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
 
     NamedIntrinsic intrinsicId = node->GetHWIntrinsicId();
 
+    if (intrinsicId == NI_AVX512F_And || intrinsicId == NI_AVX512F_AndNot || intrinsicId == NI_AVX512F_Or ||
+        intrinsicId == NI_AVX512F_Xor || intrinsicId == NI_AVX512F_ExtractVector256 ||
+        intrinsicId == NI_AVX512F_InsertVector256)
```

Comment:
This should probably also be handling `NI_AVX512F_TernaryLogic` and `NI_AVX512F_VL_TernaryLogic` (which just have `vpternlogd` and `vpternlogq`).

This otherwise looks to cover all the relevant scenarios for AVX512.

----------------

What I don't think this is handling, however, is the same scenario for `Vector128`/`Vector256`.

Consider, for example, if this were `Sse2.And` where `TYP_BYTE` is set to `INS_pand`, but the embedded mask and embedded broadcast checks will be incorrectly thinking the operand size is `1` when it should be `4` instead.

I think the most robust way to handle that would be to always normalize those nodes to have a `TYP_INT`/`TYP_UINT` `simdBaseType` on creation (regardless of whether EVEX encoding is supported). That will ensure no other phase can see an incorrect view.

We should be able to define some flag like `HW_Flag_NormalizeSmallTypeToInt` which controls when that normalization occurs and which can be used to `assert` that such nodes aren't accidentally created with small base types.

Reply:
Made the changes accordingly, the following nodes are flagged as `HW_Flag_NormalizeSmallTypeToInt `

```
SSE2/Avx2/Avx512F.And/AndN/Or/Xor
Avx512F.Insert/ExtractVector256
Avx512F/Avx512F_VL.TernaryLogic
```

importer will normalize the base type of those nodes with small base type (BYTE/UBYTE/SHORT/USHORT) to INT/UINT before actually creating the nodes.